From ca30c311a7b7d13e2e1621eff703cdc49724521d Mon Sep 17 00:00:00 2001
From: Thomas Orozco <thomas@orozco.fr>
Date: Fri, 1 Apr 2016 11:23:38 +0200
Subject: [PATCH] Panic and dump stacks on housekeeping time out

We'd rather have a process manager restart the entire cAdvisor process
rather than stop reporting altogether for a given container (which is
somewhat hard to debug otherwise).
---
 manager/container.go | 25 +++++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/manager/container.go b/manager/container.go
index 531604b..d5f4bb4 100644
--- a/manager/container.go
+++ b/manager/container.go
@@ -27,6 +27,8 @@ import (
 	"strings"
 	"sync"
 	"time"
+	"runtime/pprof"
+	"os"
 
 	"github.com/google/cadvisor/cache/memory"
 	"github.com/google/cadvisor/collector"
@@ -433,11 +435,26 @@ func (c *containerData) housekeeping() {
 }
 
 func (c *containerData) housekeepingTick() {
-	err := c.updateStats()
-	if err != nil {
-		if c.allowErrorLogging() {
-			glog.Infof("Failed to update stats for container \"%s\": %s", c.info.Name, err)
+	done := make(chan bool, 1)
+
+	go func() {
+		err := c.updateStats()
+		if err != nil {
+			if c.allowErrorLogging() {
+				glog.Infof("Failed to update stats for container \"%s\": %s", c.info.Name, err)
+			}
 		}
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		// Nothing to do
+	case <-time.After(*HousekeepingInterval * 2):
+		// We timed out. Dump all goroutine stacks to facilitate troubleshooting, and panic.
+		glog.Errorf("Housekeeping timed out for container %s", c.info.Name)
+		pprof.Lookup("goroutine").WriteTo(os.Stderr, 1)
+		panic("Aborting!")
 	}
 }
 
-- 
2.7.4

